parsing/parser.mly

```diff
@@ -2942,9 +2942,16 @@ type_variable:
 ;
 
 type_variance:
-    /* empty */                                 { Invariant }
-  | PLUS                                        { Covariant }
-  | MINUS                                       { Contravariant }
+    /* empty */                             { NoVariance, NoInjectivity }
+  | PLUS                                    { Covariant, NoInjectivity }
+  | MINUS                                   { Contravariant, NoInjectivity }
+  | BANG                                    { NoVariance, Injective }
+  | PLUS BANG                               { Covariant, Injective }
+  | MINUS BANG                              { Contravariant, Injective }
+  | INFIXOP2
+      { if $1 = "+!" then Covariant, Injective else
+        if $1 = "-!" then Contravariant, Injective else
```

Comment:
The order in the syntax is *variance*, *injectivity* (i.e. `+!` rather than `!+`), but the order in error messages is *injectivity*, *variance* --- for example, we say:

```
Error: In this definition, expected parameter variances are not satisfied.
       The 1st type parameter was expected to be injective invariant,
       but it is unrestricted.
```

Let's either:
1. allow either order in the syntax, or
2. keep the order consistent
1. allow either order in the syntax, or
2. keep the order consistent

Reply:
I agree that we should allow either order in the syntax.

Reply:
Well, this is a pain, but if you insist.

Reply:
I think that an easy approach would be to allow any number of variance and injectivity flags (in any order), and then take the intersection for each category. (You could also fail if several flags of the same category are given, but I don't think this is much better.)

Reply:
Extended parsing as suggested (and fixed the manual too).

Reply:
> I think that an easy approach would be to allow any number of variance and injectivity flags 

It would certainly be nice to be able to write:
```ocaml
type +-'a t
```
but perhaps that's better handled in a separate PR.

Reply:
I've done it in the standard way (I.e., both orders). Of course we could also support `+-` for bivariant, but then this requires more internal changes. I can do it in a separate PR if you wish.

Reply:
The relaxed order looks great.

I'd be in favour of a bivariance-annotation PR, assuming it's not a huge amount of work.

Reply:
Note: the way you have to deal with INFIXOP2 and PREFIXOP there is not nice (in particular, from the point of view of the grammar (without the failures in semantic actions we have an over-approximation of the actual language). I don't think anything pragmatic can be done in the context of the present PR (let's keep the code as you wrote it), but I think that a medium-term solution would be to stop grouping infix/prefix characters in the lexer, and instead define operator classes in the parser.

Reply:
(The suggestion to do this in the parser is hard to combine with the fact that currently spacing matters to distinguish `+ !` from `+!`, only the latter being a valid infix operator. I think it can be solved, but it requires kludges in the lexer that may not be *much* better than what we have now. Worth a try, but not high-priority.)